File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.py
def solution(numbers):
    result = [-1] * len(numbers)
    stack = []
    
    for i in range(len(numbers) - 1, -1, -1):
        current = numbers[i]
        while stack and stack[-1] <= current:
            stack.pop()
        
        if stack:
            result[i] = stack[-1]
        
        stack.append(current)
    return result

File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.js
function solution(numbers) {
    const result = new Array(numbers.length).fill(-1);
    const stack = [];
    
    for(let i = numbers.length - 1; i >= 0 ; i--){
        const current = numbers[i];
        
        while(stack.length > 0 && stack[stack.length - 1] <= current) stack.pop();
        
        if(stack.length > 0){
            result[i] = stack[stack.length - 1];
        }
        
        stack.push(current);
    }
    return result;
}